modules/imapauth/lib/Auth/Source/MyAuth.php

```diff
@@ -0,0 +1,173 @@
+			throw new Exception('Missing or invalid dsn option in config.');
+		}
+		$this->dsn = $config['dsn'];
+
+		if (!is_string($config['table_name'])) {
+			throw new Exception('Missing or invalid table_name option in config.');
+		}
+		$this->table_name = $config['table_name'];
+
+		if (!is_string($config['mail_host'])) {
+			throw new Exception('Missing or invalid mail_host option in config.');
+		}
+		$this->mail_host = $config['mail_host'];
+
+		if (!is_string($config['username'])) {
+			throw new Exception('Missing or invalid username option in config.');
+		}
+		$this->username = $config['username'];
+
+		if (!is_string($config['password'])) {
+			throw new Exception('Missing or invalid password option in config.');
+		}
+		$this->password = $config['password'];
+
+		if (!is_string($config['imap_hostname'])) {
+			throw new Exception('Missing or invalid imap_hostname option in config.');
+		}
+		$this->imap_hostname = $config['imap_hostname'];
+
+		if (!is_string($config['imap_port'])) {
+			throw new Exception('Missing or invalid imap_port option in config.');
+		}
+		$this->imap_port = $config['imap_port'];
+
+		if (!is_string($config['imap_security'])) {
+			throw new Exception('Missing or invalid imap_security option in config.');
+		}
+		$this->imap_security = $config['imap_security'];
+
+		if (is_string($config['imap_additional_options'])) {
+		$this->imap_additional_options = $config['imap_additional_options'];
+		}
+	}
+
+	protected function login($username, $password) {
+
+	//Defaults if there is no database entry
+	
+	$email=$username . "@" . $this->mail_host;
+	$email_imap = "";
+	$name=$username;
+	$uid=$username;
+
+	$imap = $this->imap_hostname . ":" . $this->imap_port . "/" . $this->imap_security;
+	if (isset($this->imap_additional_options)) $imap= $imap . $this->imap_additional_options;
+	
+	$imap = "{" . $imap . "}";
+
+	$matches = array();
+	$stripbr = array();
+
+	$sess = imap_open($imap,$username,$password,OP_READONLY,1); 
+
+		if ($sess === false ) {
+			throw new SimpleSAML_Error_Error('WRONGUSERPASS');
+		}
+	else
+	{
+	if(($last=imap_num_msg($sess))>0)
+		{
+		$head1=imap_fetchheader($sess,$last);
+		preg_match('/Envelope\-to\:(.+)/m',$head1,$matches);
+		$candidate=trim(($matches[1]));
+			if(filter_var($candidate,FILTER_VALIDATE_EMAIL)) $email_imap=$candidate;
+		}
+	elseif (imap_reopen($sess,$imap."Sent",OP_READONLY))
+		{
+		if(($last=imap_num_msg($sess))>0)
+		{
+		$head2=imap_fetchheader($sess,$last);
+		preg_match('/From\:(.+)/m',$head2,$matches);
+		$candidate=trim(($matches[1]));
+		if(preg_match('/(?<=[<\[]).*?(?=[>\]]$)/',$candidate,$stripbr))
+			{
+				$candidate=$stripbr[0];
+				if(filter_var($candidate,FILTER_VALIDATE_EMAIL)) $email_imap=$candidate;
+				}
+		}
+		}
+imap_close($sess); 
+	}
+
+	$uid=$username . "_" . $this->mail_host;
+	$user_id=$username . "_" . $this->mail_host;
+	$name=$username;
+
+	if ($email_imap != "") $email=$email_imap;
+
+if ($this->use_rc_database == true)
```

Comment:
I'm assuming this is to retrieve attributes for the user that you can't get via IMAP, right?

In that case, would it make sense to have this as a separate authentication processing filter, instead of the auth source? That way, those only interested in authentication could just use this, auth source, and if they want additional attributes they can use the authproc filter. And instead of using a configuration option for this (`use_rc_database`), they would just add the appropriate authproc filter if they want to use it.

Does it sound reasonable?